server.go

```diff
@@ -352,27 +364,78 @@ func serveFaucet(c *gin.Context) {
 		return
 	}
 
-	// Ensure the user isn't spamming the faucet
-	if user.MostRecentFaucetGrant.Add(env.FaucetRateLimit).After(time.Now()) {
-		c.JSON(http.StatusForbidden, gin.H{"error": fmt.Sprintf("you may only use the faucet once every %v hours", env.FaucetRateLimit.Hours())})
+	// No account less than a week old is allowed any FIL
+	if !user.HasAccountOlderThan(24 * time.Hour) {
+		c.JSON(http.StatusForbidden, gin.H{"error": ErrUserTooNew.Error()})
 		return
 	}
 
-	balance, err := lotusCheckBalance(ctx, targetAddr)
-	if err != nil {
+	// Check if the user is a miner
+	minerPower, minerAddr, err := lotusGetMinerInfo(ctx, targetAddr)
+	if err != nil && !strings.Contains(err.Error(), "not found") {
 		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
 		return
 	}
+	isMiner := minerPower != nil
+
+	// If not, don't let them request FIL more than once
+	if !isMiner && !user.MostRecentFaucetGrant.IsZero() {
+		c.JSON(http.StatusForbidden, gin.H{"error": "non-miners can only request FIL from the faucet once"})
+		return
+	}
 
-	owed := types.BigSub(types.BigInt(env.MaxAllowanceFIL), types.BigInt(balance))
-	dif := types.BigCmp(owed, types.NewInt(0))
+	previousAddr, err := address.NewFromString(user.FilecoinAddress)
+	if err != nil {
+		// TODO
+		// TODO
+		// TODO
+		// TODO
+		// TODO
+		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
+		return
+	}
 
-	if dif == 0 || dif == -1 {
-		c.JSON(http.StatusForbidden, gin.H{"error": "Address already has Filecoin."})
+	previousAddrMinerPower, _, err := lotusGetMinerInfo(ctx, previousAddr)
+	if err != nil && !strings.Contains(err.Error(), "not found") {
+		c.JSON(http.StatusInternalServerError, gin.H{"error": err.Error()})
 		return
 	}
 
-	cid, err := lotusSendFIL(ctx, faucetAddr, targetAddr, types.FIL(owed))
+	var owed types.BigInt
+	if isMiner {
+		// Ensure the user isn't spamming the faucet: no user can request FIL more
+		// than once every 24 hours unless they've *just* become a miner
+		wasAlreadyMiner := previousAddrMinerPower != nil
+		justBecameMiner := isMiner && !wasAlreadyMiner
+		requestIsTooSoon := user.MostRecentFaucetGrant.Add(env.FaucetRateLimit).After(time.Now())
+		if !justBecameMiner && requestIsTooSoon {
+			c.JSON(http.StatusForbidden, gin.H{"error": fmt.Sprintf("you may only use the faucet once every %v hours", env.FaucetRateLimit.Hours())})
+			return
+		}
+
+		// Calculate what the user is owed
+		//
+		//     [min: 2000, max: (400 * miner_size_TiB / elapsed_days)]
+		//
+		daysSinceLastGrant := uint64(time.Now().Sub(user.MostRecentFaucetGrant).Hours()) / 24
+		if daysSinceLastGrant == 0 {
+			daysSinceLastGrant = 1
+		}
```

Comment:
Not sure I understand the logic for this. If we follow the approach I outlined in the main thread, we should just be able to just if `requestIsTooSoon` and then `daysSinceLastGrant` will never be 0?